jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+                    }
+                }
+            }
+
+            if (!upgrade)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no upgrade header {}",baseRequest);
+                return false;
+            }
+
+            if (connectionCSVs==null || !connectionCSVs.getValues().contains("upgrade"))
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection upgrade {}",baseRequest);
+                return false;
+            }
+
+            if (key==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no key {}",baseRequest);
+                return false;
+            }
+
+            // Create a connection from a connection factory (from context or connector)
+            ServletContext context = baseRequest.getServletContext();
+            WebSocketConnectionFactory connectionFactory = null;
+            if (context!=null)
+                connectionFactory = (WebSocketConnectionFactory)context.getAttribute(WebSocketConnectionFactory.class.getName());
+            if (connectionFactory==null)
+                connectionFactory = baseRequest.getHttpChannel().getConnector().getBean(WebSocketConnectionFactory.class);
+            if (connectionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection factory {}",baseRequest);
+                return false;
+            }
+            HttpChannel channel = baseRequest.getHttpChannel();
+            WebSocketCoreConnection connection =
+                    connectionFactory.newConnection(channel.getConnector(),channel.getEndPoint(),extensionCSVs.getValues());
+            if (connection==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no connection {}",baseRequest);
+            }
+
+            // Create a session from a session factory (from context or connector)
+            WebSocketSessionFactory sessionFactory = null;
+            if (context!=null)
+                sessionFactory = (WebSocketSessionFactory)context.getAttribute(WebSocketSessionFactory.class.getName());
+            if (sessionFactory==null)
+                sessionFactory = baseRequest.getHttpChannel().getConnector().getBean(WebSocketSessionFactory.class);
+            if (sessionFactory==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session factory {}",baseRequest);
+                return false;
+            }
+            WebSocketCoreSession session = sessionFactory.newSession(connection, subprotocol);
+            if (session==null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("not upgraded no session {}",baseRequest);
+            }
+
+            connection.setSession(session);
```

Comment:
This is incorrect.

The `session.setWebSocketEndpoint(Object endpoint, WebSocketPolicy policy, L localEndpoint, R remoteEndpoint)` needs to be called instead.

Reply:
While the connection needs access to the session, the way the other 4 APIs are doing this now is via the creation of the connection (which needs a session), extend WebSocketCoreConnection and have its constructor call .setSession(connection) instead.

Reply:
The connection uses the session for ...

* The destination of parsed frames (after the extensionstack) `session.incomingFrames()`
* Reporting the Jetty `Connection.onOpen()` to the RFC6455 OPEN phase `session.open()`, which triggers the `localEndpoint.onOpen()`
* Reporting Connection level errors via the `session.processError(Throwable)`

To eliminate this ...

* Using `Connection.Listener.onOpened()` can be used for the `session.open()`
* Passing `IncomingFrames` instead of Session to the connection perhaps like `connection.setIncoming(session)`
* Using an new interface such as `ConnectionErrorListener.onConnectionError(Throwable t)` which the session implements instead.

Reply:
The `session.setWebSocketEndpoint(Object endpoint, WebSocketPolicy policy, L localEndpoint, R remoteEndpoint)` is done in my test code within the session factory.   The core code doesn't need to know how to construct a session, it only needs to link the abstract session to the connection, which is exactly what `connection.setSession(session)` does.